src/samplers/cpu/linux/usage/bpf.rs

```diff
@@ -142,34 +163,34 @@ impl CpuUsage {
         // refresh the counters from the kernel-space counters
         self.bpf.refresh_counters(elapsed);
 
-        // get the new sum of all the counters
-        let sum_now: u64 = sum();
-
-        // get the number of nanoseconds in busy time, since idle hasn't been
-        // incremented, the busy time is the difference between our prev and
-        // current sums
-        let busy_delta = sum_now.wrapping_sub(self.sum_prev);
+        // update busy time metric
+        let busy: u64 = busy();
+        let busy_prev = CPU_USAGE_BUSY.value();
+        let busy_delta = busy.wrapping_sub(busy_prev);
+        self.total_busy.set(elapsed.as_secs_f64(), busy);
 
-        // idle time delta is `cores * elapsed - busy_delta`
-        let idle_delta = self.online_cores as u64 * elapsed.as_nanos() as u64 - busy_delta;
-
-        // update the idle time metrics
-        CPU_USAGE_IDLE.add(idle_delta);
-        let _ = CPU_USAGE_IDLE_HISTOGRAM.increment(idle_delta);
+        // calculate the idle time elapsed since last sample, update metric
+        let idle_prev = CPU_USAGE_IDLE.value();
+        let idle_delta =
+            (self.online_cores as u64 * elapsed.as_nanos() as u64).saturating_sub(busy_delta);
+        self.total_idle
+            .set(elapsed.as_secs_f64(), idle_prev.wrapping_add(idle_delta));
 
         // do the same for percpu counters
-        for (cpu, sum_prev) in self.percpu_sum_prev.iter_mut().enumerate() {
-            let sum_now: u64 = self.percpu_counters.sum(cpu).unwrap_or(0);
-            let busy_delta = sum_now.wrapping_sub(*sum_prev);
-            let idle_delta = elapsed.as_nanos() as u64 - busy_delta;
-            self.percpu_counters
-                .add(cpu, IDLE_CPUTIME_INDEX, idle_delta);
-            *sum_prev += busy_delta + idle_delta;
+        for (cpu, (busy_counter, idle_counter)) in self
+            .percpu_busy
+            .iter_mut()
+            .zip(self.percpu_idle.iter_mut())
+            .enumerate()
+        {
+            let busy: u64 = self.percpu_counters.sum(cpu).unwrap_or(0);
+            let busy_prev = busy_counter.set(busy);
+            let busy_delta = busy.wrapping_sub(busy_prev);
+
+            let idle_delta = (elapsed.as_nanos() as u64).saturating_sub(busy_delta);
```

Comment:
Shouldn't this check if the core is online? If we don't have that information, it seems like delta and idle will be perfect duals of each other and we can just track either one.

Reply:
Yes. This is fixed now.